### ISSUE TITLE
MemoryError

### ISSUE THREAD

**Issue by drtianfenghao** [pgmpy/pgmpy#1177]
### Subject of the issue
I have  creare a BN.
model = BayesianModel([
                ("A","SD"),("DW","SD"),
                ('A', 'ES'), ('DW', 'ES'),('SD','ES'),
                ('IW', 'IE'), ('DNR', 'IE'),('G','IE'),
                ('DRW', 'LC'), ('DT', 'LC'), ('DRD', 'LC'),
                ('DW', 'C'),("LC","C"),
                ('ES', 'ELP'), ('IE', 'ELP'),("C","ELP"),
                ("ELP","STKJ"),("ELP","SCKJ"),("ELP","SHKJ")
                ])
But,it has some issue with "MemoryError" when "print (student_infer.map_query(variables=['SCKJ','STKJ','SHKJ'],evidence={'ELP':1}))"

How deal this?

### Your environment
* pgmpy 2.7
* Python 2.9
* Operating System Windows10

### Steps to reproduce
data = pd.DataFrame(df,columns=df.columns)
data.head()
model = BayesianModel([
                ("A","SD"),("DW","SD"),
                ('A', 'ES'), ('DW', 'ES'),('SD','ES'),
                ('IW', 'IE'), ('DNR', 'IE'),('G','IE'),
                ('DRW', 'LC'), ('DT', 'LC'), ('DRD', 'LC'),
                ('DW', 'C'),("LC","C"),
                ('ES', 'ELP'), ('IE', 'ELP'),("C","ELP"),
                ("ELP","STKJ"),("ELP","SCKJ"),("ELP","SHKJ")
                ])
model.fit(data, estimator=MaximumLikelihoodEstimator)
student_infer = VariableElimination(model)
print (student_infer.map_query(variables=['SCKJ','STKJ','SHKJ'],evidence={'ELP':1}))

### Expected behaviour
Tell us what should happen

### Actual behaviour
Tell us what happens instead

**Comment by ankurankan** [comment]
@drtianfenghao Could you give me an idea of the number of states for the variables in the model? Since the model isn't that big, I think the MemoryError might be because of a large number of states for the variables.

**Comment by drtianfenghao** [comment]
Thank you for your attention！The data in attachment
my code:
model = BayesianModel([
                ("A","SD"),("DW","SD"),
                ('A', 'ES'), ('DW', 'ES'),('SD','ES'),
                ('IW', 'IE'), ('DNR', 'IE'),('G','IE'),
                ('DRW', 'LC'), ('DT', 'LC'), ('DRD', 'LC'),
                ('DW', 'C'),("LC","C"),
                ('ES', 'ELP'), ('IE', 'ELP'),("C","ELP"),
                ("ELP","STKJ"),("ELP","SCKJ"),("ELP","SHKJ")
                ])model.fit(data, estimator=MaximumLikelihoodEstimator)

infer = VariableElimination(model)

print infer.map_query(variables=['STKJ','SCKJ','SHKJ'],evidence={'ELP':1})
error:
phi.values = phi.values * phi1.values
MemoryError





------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Ankur Ankan"<notifications@github.com&gt;;
发送时间:&nbsp;2019年10月29日(星期二) 晚上9:32
收件人:&nbsp;"pgmpy/pgmpy"<pgmpy@noreply.github.com&gt;;
抄送:&nbsp;"田丰昊"<593088840@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [pgmpy/pgmpy] MemoryError (#1177)




@drtianfenghao Could you give me an idea of the number of states for the variables in the model? Since the model isn't that big, I think the MemoryError might be because of a large number of states for the variables.

**Comment by ankurankan** [comment]
@drtianfenghao I can't see the data. I don't think github allows to attach data to comments. Could you either email it to me or share a link?

**Comment by ChuWen94** [comment]
@ankurankan, I would be interested if this issue has gotten resolved. I've had a similar issue - the model works fine until the number of nodes get increased.

**Comment by ankurankan** [comment]
@ChuWen94 I am working ( #1239 ) on some optimizations based on structure pruning based on the query. This should be able to decrease the memory usage of inference algorithms in sparse structures.

**Comment by ChuWen94** [comment]
@ankurankan awesome - thank you for the update!

**Comment by SrGrace** [comment]
@ankurankan What is the limit for BayesianModel or any other model of pgmpy in terms of Nodes, Edges, level of the graphs and number of states etc?

**Comment by ankurankan** [comment]
@SrGrace It's very hard to give a number as all the factors (nodes, edges, no. of states) affect each other (non-linearly) in algorithms. And in this specific inference case, it is also quite dependent on how efficient the `elimination_order` is. If you are getting a memory error, I would suggest the following steps in order:
1. See if you can think of an efficient elimination order. If you have one, specify `elimination_order` instead of using a heuristic to compute it.
2. If there are some variables in the network which have a lot of parents, try to remove some of the parents.
3. Try to reduce the number of states of some variables.

**Comment by ivanDonadello** [comment]
Dear @ankurankan, I also have a memory error. My network has 22 nodes, 21 edges, all the nodes have 5 states/values except one that has only 2 states/values. The values are discrete. I have memory problems (both on my desktop with 32Gigas and on Colab) with this simple query:

```
variables = ['18', '19']
evidence = {'age': 1 , 'sex': 0}
values = pd.DataFrame(np.array([[1, 0],[4, 0],[1, 1],[4, 1]]), columns=['age', 'sex'])
prediction = model.predict(values, n_jobs=-1)
```

I also tried with n_jobs=6. I don't have variables with many parents. How can I solve this? Here the error:

```
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 188.59it/s]
*** joblib.externals.loky.process_executor.TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

The exit codes of the workers are {SIGKILL(-9)}
`
``

Thanks for your attention

**Comment by ankurankan** [comment]
@ivanDonadello In the example that you have shown, the problem is that `values` is defined on only 2 variables, hence the predict method will try to compute a joint probability distribution (and return the most likely state) on the rest 20 variables which will have 5^20 values, and hence the memory error. I am not sure if you actually wanted to predict 20 variables or were expecting something else from the method?

**Comment by ivanDonadello** [comment]
@ankurankan thanks for the answer. I tried different numbers of variables for the evidence and the minimum working number of evidence variables is 11. Actually, I need to predict 17 variables with an evidence of 5 variables. Why? This evidence comes from user answers to a questionnaire therefore I cannot pose all 22 questions as evidence. So my problem is how to select the smallest number of variables to use as evidence that gives me quite good predictions.

**Comment by ankurankan** [comment]
@ivanDonadello I am sorry but I am not able to understand your usecase. Could you possibly give me more details? I don't think it would be possible to predict 17 variables together with a Bayesian Network. To do inference on 17 variables together (with 5 states each) would need at least 1 TB of memory and quite a lot of computation time. And if you care just about the prediction accuracy (and not model the data generating distribution), I think other machine learning methods might be able to give better results.

**Comment by ivanDonadello** [comment]
@ankurankan thanks for your answer, I think that maybe another ML algorithm would better fit my usecase.

To answer your question. I have 22 variables about demographic information of users and their preferences. 5 of them are evidence (sex, age, school, etc), the other 17 are preference scores (I like fish, I could try white meat instead of red one, I like legumes) from 0 to 4. Once I have trained a bayesian network on this data my idea would be: can I infer the 17 variables from only the 5 demographic info? Should I need to ask more evidence? Maybe I can pose some questions from the 17 preferences to get more evidence. However, how do I choose the other questions?

I hope this clarifies.

**Comment by MathijsPost** [comment]
> 
> 
> @SrGrace It's very hard to give a number as all the factors (nodes, edges, no. of states) affect each other (non-linearly) in algorithms. And in this specific inference case, it is also quite dependent on how efficient the `elimination_order` is. If you are getting a memory error, I would suggest the following steps in order:
> 
>     1. See if you can think of an efficient elimination order. If you have one, specify `elimination_order` instead of using a heuristic to compute it.
> 
>     2. If there are some variables in the network which have a lot of parents, try to remove some of the parents.
> 
>     3. Try to reduce the number of states of some variables.

Can I ask an additional question regarding variable elimination? I'm working with some random samples and often the query returns NaN values. When or why would this happen?

**Comment by ankurankan** [comment]
@MathijsPost I can't think of a reason why that might happen. Do you by any chance have a reproducible code that I can look at?

**Comment by MathijsPost** [comment]
Hi @ankurankan I looked into it, I was creating random samples for every variable, which led to combinations of variable states that never happened in my case. For example, I have a variable that includes the hour of the day (0-23) and another variable that indicates the moment of the day (morning, afternoon, evening, night). There were combinations that said it was 07:00 AM for hour of the day and 'evening' for moment of the day, which makes no sense, never occurred in the data and hence resulted in NaN.

I solved this by finding all unique combinations of the variables I am interested in.

**Comment by MathijsPost** [comment]
Hi @ankurankan , just a quick follow-up. If I enter a query with a certain combination of evidence variables that never occurred in the data on which the model is fitted, say I change one variable in the evidence to test sensitivity, can pgmpy handle this? It seems that the query returns NaN values.

**Comment by ankurankan** [comment]
@MathijsPost Yeah, I don't think it can handle such situations right now. But one workaround to this could be to manually specify the state names (including the ones that do not occur in the dataset) while learning the parameters: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/estimators/MLE.py#L26

**Comment by MathijsPost** [comment]
Hi @ankurankan, all state names do occur in the dataset, it is just that the specific combination of evidence never occurred in the dataset. Or should this be done for every combination?

**Comment by ankurankan** [comment]
@MathijsPost Do you by any chance have a reproducible code, so that I can try to debug it? My best guess is that because of some missing combination of states, some of the CPDs (which are on this combination of variables) have that specific combination of states missing, which might be resulting in a NaN. I think providing the `state_names` manually might also be able to fix it.

**Comment by MathijsPost** [comment]
Here is the code:
```
import random
import pandas as pd
from pgmpy.inference import VariableElimination
from pgmpy.readwrite import BIFReader

trainingdata = pd.read_pickle('trainingdatav10.pkl')

def loadmodel(name):
    reader = BIFReader(name + '.bif')
    loadedmodel = reader.get_model(state_name_type=int)
    return loadedmodel

#%%
structurelearningmodel = loadmodel("2021-04-10_12_37_structure_2021-04-10_09_19_k2score100000_100max_parents_7")
#%%
inference = VariableElimination(structurelearningmodel)

parameters = ['inbound_peak_num', 'day_moment', 'met_condition', 'configtype', 'day_of_the_week', 'month_num', 
              'heavies_in_the_mix', 'arr_congestion_index_20_SCH', 'arrival_congestion_index',
             'number_of_arrivals_daily', 'cumulative_delay_since_first_peak',
             'cumulative_delayed_flights_since_first_peak', 'declared_rate', 'config'] 

'''
Add parameter of interest:
'Airline', 'Continent', 'FIR_delay', 'arrival_congestion_index', 'declared_rate', 'dep_airp', 'heavies_in_the_mix', 
'landing_RWY', 'schedule_buffer_estimate', 'taxi_time', 'wtc', 'number_of_arrivals_daily', 'cumulative_delay_since_first_peak',
'cumulative_delayed_flights_since_first_peak', 'met_condition'
'''

parameter_of_interest = 'declared_rate'
if parameter_of_interest not in parameters:
    parameters.append(parameter_of_interest)

parametercombinations = pd.DataFrame()
parametercombinations['lists'] = trainingdata[parameters].values.tolist()
parametercombinations['lists_str'] = parametercombinations.lists.astype(str)
parametercombinations = parametercombinations.drop_duplicates('lists_str')
parametercombinations = parametercombinations.lists.tolist()

#%% 
conditions2 = pd.DataFrame()
indexlist2 = parameters.copy()
indexlist2.extend(['ATFM_aerodrome_cap_delay', 'Delay_min'])
conditions2['indexlist'] = indexlist2
conditions2 = conditions2.set_index('indexlist')
a = 0
for i in range(0,5):
    print('----- Inference', i, '-----')
    evidencelist2 = random.choice(parametercombinations)
    evidencedict2 = dict(zip(parameters, [int(i) for i in evidencelist2]))

#this is where I am trying to change the value of one variable, the parameter_of_interest, to see the effects of this one variable
    for j in range(0, structurelearningmodel.get_cardinality(parameter_of_interest)):
        print('----- Parameter value', j, '-----')
        evidencedict2[parameter_of_interest] = j
        phi2 = inference.query(['Delay_min', 'ATFM_aerodrome_cap_delay'], 
                              evidence = evidencedict2, joint=False, 
                              elimination_order = 'MinFill')
                
        values2 = list(evidencedict2.values())
        values2.append(phi2['ATFM_aerodrome_cap_delay'].values)
        values2.append(phi2['Delay_min'].values)
        conditions2['%d' % a] = values2
        a+=1
```

I'll send you the .pkl and .bif file via e-mail.

**Comment by MathijsPost** [comment]
@ankurankan one addition: it doesn't always return NaN when I enter an evidence combination that did not occur in the data

**Comment by jalalmahmud** [comment]
We have been trying pgmpy for inference using bayesian network. For these configurations, getting OOM errors. I can realize that these are very huge networks, so inference is computationally expensive. My question is whether it is reasonable to expect current pgmpy inference algorithm (e.g., variable elimination) to work for such large networks? 

# nodes: 340
# edges: 1111
# states: 31

# nodes: 81
# edges: 150
# states: 16

# nodes: 453
# edges: 1224
# states: 31

**Comment by ankurankan** [comment]
@jalalmahmud I don't think such large networks would work. A couple of ways you can reduce complexity is by somehow reducing the number of parents of variables (i.e., reduce the max. indegree) or reduce the number of states of these variables.

**Comment by jalalmahmud** [comment]
Thank you for your response, and indicating that such large network is not feasible. My understanding is that most costly part of inference is joint probability computation. Perhaps, some approximation could be done to include only the parents of a node (or may be the nodes in markov blanket) for computing joint probability computation? 

I was looking at the example in this page: 

https://pgmpy.org/detailed_notebooks/2.%20Bayesian%20Networks.html

So following the example in this page, to compute P(G), perhaps we could only compute P(G) in terms of P(I) and P(D)? [i.e. parents of G], then the compute would be much simpler? 

So, P(G) = P(I) * P (D) * P(G | D, I) 

Would that be an approximation to reduce compute?

**Comment by ankurankan** [comment]
@jalalmahmud In that specific example, using only G and I does work but in the general case considering only the parents would give incorrect results as all ancestors need to be considered. We already have optimizations (https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/inference/base.py#L120) that for a given query only considers the relevant nodes and discards everything else. 

I think in your case, the main issue is the number of states of the variables. Even with the approximate methods like: https://pgmpy.org/approx_infer/bn_sampling.html, you would end up getting 0 probability for most of the queries.Maybe a Gibbs sampling kind of approach might work: https://pgmpy.org/approx_infer/gibbs.html but with the current implementation, we can only compute marginal distributions.

**Comment by jalalmahmud** [comment]
Thanks again for your response, and pointing to the code where pruning is done to optimize computations. I have some follow-up questions based on your prior response. So if pruning is already in place during query, then what is the main reason for the limit in terms of how many nodes/edges the original network can be? [Your earlier response indicates we shouldn't probably aim to build large network with 100+ nodes and edges]. 

Also, about limiting the states, I understand that even if we prune a network, too many states significantly increases compute/memory. We tried something like this before: 

# nodes: 154
# edges: 454
# states: 5

However, this one also failed during inference. So perhaps here nodes/edges should be also pruned in the first place?

**Comment by ankurankan** [comment]
@jalalmahmud The limit is decided by the maximum size of the intermediate matrices. As the size of these matrices is a function of the number of parents of the nodes and the number of states of these nodes, both of these together determine the limit. Essentially, we need to keep the max indegree low (i.e., there should not be any node with a lot of parents), and the number of states should be low. There are also techniques to reduce the max indegree of the model by inserting latent variables in the model (called parent divorcing), but is unfortunately not implemented in pgmpy yet: https://github.com/pgmpy/pgmpy/issues/1814

**Comment by jalalmahmud** [comment]
> @jalalmahmud In that specific example, using only G and I does work but in the general case considering only the parents would give incorrect results as all ancestors need to be considered. We already have optimizations (https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/inference/base.py#L120) that for a given query only considers the relevant nodes and discards everything else.
> 
> I think in your case, the main issue is the number of states of the variables. Even with the approximate methods like: https://pgmpy.org/approx_infer/bn_sampling.html, you would end up getting 0 probability for most of the queries.Maybe a Gibbs sampling kind of approach might work: https://pgmpy.org/approx_infer/gibbs.html but with the current implementation, we can only compute marginal distributions.

Regarding parents/ancestors, can't we just use parents as per this formula? 

the joint distribution for a Bayesian network is equal to the product of P(node|parents(node)) for all nodes, stated below:

![image](https://github.com/user-attachments/assets/ad286767-9878-4777-84af-ff57e841b84b)

**Comment by ankurankan** [comment]
@jalalmahmud No, this dependence on just the parents only allows us to factorize the model's parameterization, i.e., we can define CPDs of a node given its parents. But when doing inference we need to consider all nodes as there is a product over all the nodes in the formula above.

**Comment by jalalmahmud** [comment]
Thanks, regarding product over all the nodes, 
If I try the above formula for n = 2, then P(X1, X2) = P(X1|Parent(X1)) P (X2|Parent(X2)) 
                                           for n = 3, P(X1, X2, X3) = P(X1|Parent(X1)) P (X2|Parent(X2)) P (X3|Parent(X3))
What am I missing here?

**Comment by ankurankan** [comment]
@jalalmahmud In most cases, you can't compute the product of distributions like that, it will very quickly become intractable as the size of the joint distribution is exponential in the number of variables. Inference algorithms are designed to exactly avoid this full product.
